### ISSUE TITLE
When using an async preValidation hook the handler will be called multiple times

### ISSUE THREAD

**Issue by rijx** [fastify/fastify#1511]
## 🐛 Bug Report

When using an async preValidation hook the handler will be called multiple times.

## To Reproduce

Run this code with fastify 2.0.0 installed:

```js
const fastify = require("fastify")();

fastify.addHook("preValidation", async function preValidation(
  request,
  reply,
  next
) {
  await new Promise(resolve => setTimeout(resolve, 1000));

  next();
});

fastify.get("/", async (request, reply) => {
  console.log(1);

  await new Promise(resolve => setTimeout(resolve, 10));

  console.log(2);

  await new Promise(resolve => setTimeout(resolve, 10));

  reply.send("OK");
});

fastify.listen(8090);
```

Then run (or use a browser):

```
curl localhost:8090
```

## Expected behavior

The console should output:

```
1
2
```

What happens instead:

```
1
1
2
2
```

## Your Environment

- *node version*: v8.14.0
- *fastify version*: 2.0.0
- *os*: macOS

**Comment by jsumners** [comment]
You should not be passing in a `next` callback or invoking said callback.

**Comment by rijx** [comment]
You're right. I overlooked that...

It may have been better for it to be called `callback` instead of `next` if it's called implicitly for async functions.

I personally thought it functioned like Koa's next, hence the confusion.

**Comment by jsumners** [comment]
The docs provide examples for both styles. Is it not clear? https://github.com/fastify/fastify/blob/b4560139a74d5a40c58d30d1bd7dd2dc21a1b153/docs/Hooks.md#requestresponse-hooks

**Comment by rijx** [comment]
What I think what went wrong:

I first had a sync preValidation and now made it async/await. It didn’t occur to me that I needed to change the control flow to not call next anymore (do I need to call it for sync functions anyway?)

When I hit this strange issue I overlooked the differences between the sync and async examples in the docs.

I personally would check for next being called multiple times in debug mode, but you’re right that it’s documented and I wouldn’t have ran into this issue if I read the docs better.

Feel free to close the issue, I’m happy that my code works again. Cheers.
